lib/commands/release/finalize.ts

```diff
@@ -0,0 +1,79 @@
+/**
+ * @license
+ * Copyright 2016-2020 Balena Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { flags } from '@oclif/command';
+import Command from '../../command';
+import * as cf from '../../utils/common-flags';
+import { getBalenaSdk, stripIndent } from '../../utils/lazy';
+import { getRelease } from '../../utils/release-utils';
+
+interface FlagsDef {
+	help: void;
+}
+
+interface ArgsDef {
+	commitOrId?: string;
+}
+
+export default class ReleaseFinalizeCmd extends Command {
+	public static description = stripIndent`
+		Finalize a release.
+`;
+	public static examples = [
+		'$ balena release finalize commit',
+		'$ balena release finalize id',
+	];
+
+	public static usage = 'finalize <commitOrId>';
```

Comment:
Re `commitOrId`, what is a release ID? Is it the numeric database ID, like 1234567 ? I am considering #1843, although I think the web dashboard exposes the numeric database ID and maybe even the output of `balena push` exposes it at end of the build (?). I see pros and cons in accepting a database ID here, but a product decision was made in past to stop exposing and using numeric database IDs.

I understand that a "release commit" had its origins in git commits and `git push` to balenaCloud, when the release commit was actually exactly the same value as the git commit. `balena push` disregards the git commit, and a release UUID is then created and named a commit, for consistency and backwards compatibility. This explanation of mine is in part informed guessing, just to get us thinking about it. :-)

Reply:
Ah, only now I noticed outdated / hidden review comments, where @thgreasi had requested numeric database IDs to be added despite #1843, almost the opposite of my review comment:

> Maybe we could also document using release IDs, since commits are not even unique in the same fleet.
> (The commit is unique only across successful releases of a given fleet, hence you can get an AmbiguousReleaseError in case i) your first `git push` failed, or you are `git push`ing the same commit to more than one fleets.)

Non uniqueness of release commits is indeed a strong argument for accepting numeric database IDs, and make for an uneasy, incorrect "semantic mapping" between commit hashes and UUIDs, in the sense that UUIDs are properly unique.

To the extent that commit hashes are not unique while past product call decisions were made to avoid exposing numeric database IDs, in theory one could think of a feature request where the API data model should add "proper" release UUIDs in addition to release commits and numeric database IDs. How messy! 😅

Reply:
Given #1843 and how `balena push` is our suggested way of creating new releases, I think I'm fine supporting only commits (if @pdcastro insists) and then bringing up a selector for the user to select the target release. We could probably also filter out failed releases, since it doesn't offer anything to finalize them, but that would still be an issue for the `balena release` command.

Reply:
I think filtering out failed releases is the correct approach for `balena release finalize` since, as I understood, the commit is unique for final releases right? Other alternative that could work for this and `balena release` is to sort by status (with successful releases first) and return just the first result. My guess is that in most cases that would return what the user wants.

Reply:
I don't insist! 😁  Relying on human interaction (bringing up a selector for the user to select the target release) is an even more serious issue than exposing numeric database IDs in my opinion (I have only ever seen 1 user complaint about numeric database IDs as they thought it was somehow a security issue, while I have seen many complaints on difficulties to programmatically use the CLI because of colours, unicorns, non-JSON output, interactive menus, etc, e.g. #1089, for CLI use in scripts and/or CI environments).

The proper solution might actually be to add release UUIDs to the data model in addition to release commits, but that's outside the scope of this PR. On balance, I think in this case we are better off allowing numeric database IDs, as implemented.

Reply:
> what is a release ID? Is it the numeric database ID, like 1234567 ?

Correct.  These were specifically requested by @thgreasi to disambiguate releases.

Reply:
I am hesitant about this.  This would cause a different list from the CLI that the dashboard shows.  I think we need to show them all, or else add a flag `--all`.

Reply:
> I think filtering out failed releases is the correct approach

I suspect Felipe meant filtering out failed releases if the commit given as argument to `balena release finalize <commit>` was ambiguous. Say there were 3 releases with commit `a777f7345fe3d655c1c981aa642e5555`, which I think (not sure) you could try and reproduce by using the `git push` deployment method (rather than `balena push`) with a Dockerfile containing a RUN line that executes a script that fails (non zero exit code) if it is executed before midday, and succeeds after midday (like me, I work better in the afternoon). If you `git push` that git commit twice before midday, you get two failed releases. Then `git push` the same commit again after midday, and the release succeeds. Then try `balena release finalize <commit>` (by commit, not database ID). Presumably the SDK will throw some "Release is ambiguous" error. In this case, Thodoris suggest using an interactive menu for the user to select the right release to finalize, and Felipe reasoned: there can be many failed releases matching a given commit, but only one successful release with that same commit (enforced by the API schema), so we should filter out the failed releases and finalize the successful one, avoiding the need for an interactive menu and avoiding the need for user to use a database ID to disambiguate between releases that share the same commit.

 it was in reply to Thodoris' suggestion of an interactive menu for the user to disambiguate the commit in case they ran `balena release finalize a777f7345fe3d655c1c981aa642e5555` and there were multiple releases

Reply:
Sorry I had not seen Paulo's response. But what he describes is exactly what I meant, however much better explained :p

Reply:
From discussion with Thodoris and PJ, it turns out disambiguation for `finalize` is not really necessary since releases created with the `git` flow are always final. However for consistency with the `release` info (and for the sake of time), we agreed  that keeping the `commitOrId` input of both is fine for now. 

We will need to discuss how to make release `commit` an unambiguous identier on the API at some point in the future.